mypy/fastparse2.py

```diff
@@ -872,68 +891,72 @@ def visit_Num(self, new: ast27.Num) -> Expression:
         elif isinstance(value, complex):
             expr = ComplexExpr(value)
         else:
-            raise RuntimeError('num not implemented for ' + str(type(new.n)))
+            raise RuntimeError('num not implemented for ' + str(type(n.n)))
 
         if is_inverse:
             expr = UnaryExpr('-', expr)
 
-        return expr
+        return self.set_line(expr, n)
 
     # Str(string s)
-    @with_line
-    def visit_Str(self, s: ast27.Str) -> Expression:
+    def visit_Str(self, n: ast27.Str) -> Expression:
         # Hack: assume all string literals in Python 2 stubs are normal
         # strs (i.e. not unicode).  All stubs are parsed with the Python 3
         # parser, which causes unprefixed string literals to be interpreted
         # as unicode instead of bytes.  This hack is generally okay,
         # because mypy considers str literals to be compatible with
         # unicode.
-        if isinstance(s.s, bytes):
-            n = s.s
+        if isinstance(n.s, bytes):
+            s = n.s
             # The following line is a bit hacky, but is the best way to maintain
             # compatibility with how mypy currently parses the contents of bytes literals.
-            contents = str(n)[2:-1]
-            return StrExpr(contents)
+            contents = str(s)[2:-1]
+            e = StrExpr(contents)  # type: Union[StrExpr, UnicodeExpr]
+            return self.set_line(e, n)
         else:
-            return UnicodeExpr(s.s)
+            e = UnicodeExpr(n.s)
+            return self.set_line(e, n)
 
     # Ellipsis
     def visit_Ellipsis(self, n: ast27.Ellipsis) -> EllipsisExpr:
         return EllipsisExpr()
 
     # Attribute(expr value, identifier attr, expr_context ctx)
-    @with_line
-    def visit_Attribute(self, n: ast27.Attribute) -> Expression:
-        if (isinstance(n.value, ast27.Call) and
-                isinstance(n.value.func, ast27.Name) and
-                n.value.func.id == 'super'):
-            return SuperExpr(n.attr, self.visit(n.value))
-
-        return MemberExpr(self.visit(n.value), n.attr)
+    def visit_Attribute(self, n: Attribute) -> Expression:
+        member_expr = MemberExpr(self.visit(n.value), n.attr)
+        obj = member_expr.expr
+        if (isinstance(obj, CallExpr) and
+                isinstance(obj.callee, NameExpr) and
+                obj.callee.name == 'super'):
+            e = SuperExpr(member_expr.name, obj)  # type: Expression
+        else:
+            e = member_expr
+        return self.set_line(e, n)
```

Comment:
I am curious, does this (using `isinstance()` with native classes in this methid) give a visible perf win?

Also maybe it is worth adding a comment, otherwise it is not clear why we always create a `MemberExpr` instead of implementing this as it was before.

Reply:
I didn't measure the impact of each individual change, but this should be slightly faster since super expressions are much less common than member expressions, and the check for a super expression will be faster in the new implementation since we can do native attribute access (`obj.callee` and such). The impact of this particular change may not be measurable, but the overall reduction in Python attribute accesses gave a measurable aggregate performance increase in the Python 3 AST transformation case at least. Will add a comment.